### ISSUE TITLE
API: Make it possible to use pre-instantiated Java-based migrations

### ISSUE THREAD

**Issue by guylabs** [flyway/flyway#1062]
Hi,

as we are using Spring with Flyway we want to use other beans inside the `SpringJdbcMigration`s. Therefore we created a new resolver which retrieves the `SpringJdbcMigration`s from the Spring application context. This way we are able to define the migrations inside the application context and reference other beans we would need for this migration inside the `SpringJdbcMigraiton` with the dependency injection mechanisms of Spring (`@Autowired`). Please have a look at the linked pull request and the tests how this is done.

If this pull request is accepted and integrated, then I would also add some documentation to the website project of Flyway on how to use this new resolver.

Thanks and regards,

Guy

**Comment by kapil-malik** [comment]
Hi,
This is very useful for us as well. The PR looks pretty cool. Any updates on how soon can it be part of flyway builds? 
Till then we will override SpringJbdcMigrationResolver (like your code) as a workaround.

**Comment by guylabs** [comment]
@kapil-malik Thanks and yes this would be really useful. 

@axelfontaine Thanks a lot for taking it up as a feature!

**Comment by wimdeblauwe** [comment]
+1 for this!

**Comment by crstalli** [comment]
@wimdeblauwe Thank you very much! That was extremely helpful!

**Comment by marbot** [comment]
+1, that would be a very useful feature.

**Comment by axelfontaine** [comment]
See also #869

**Comment by geirgp** [comment]
+1

**Comment by tpischke** [comment]
+1

**Comment by paolodenti** [comment]
+1

**Comment by K-Yosuke** [comment]
+1

**Comment by zinefer** [comment]
When I use Spring JPA flyway migrations seem to run before the application context has been instantiated ... Does anyone know of a way around this, or how I can instantiate my repositories inside a flyway migration?

**Comment by OldIMP** [comment]
We use a workaround from https://github.com/avehlies/spring-beans-flyway2 But I really don't recommend injecting data repo in migrations since the domain could have been changed.

**Comment by zinefer** [comment]
Yes, I've tried that technique... But in my project, flyway migrations are being run before the application context - I get Null exceptions.

**Comment by zinefer** [comment]
Infact, if you take that project (spring-beans-flyway2) and add `compile("org.springframework.boot:spring-boot-starter-data-jpa")` to the dependencies in `build.gradle` it will fail with a null pointer exception

**Comment by OldIMP** [comment]
Sry I forgot to mention our configuration:
```java
@Configuration
class FlywayConfiguration {
    @Primary
    @Bean(name = "flywayInitializer")
    @DependsOn("springUtility") //or whatever your ApplicationContextAware is named
    FlywayMigrationInitializer flywayMigrationInitializer(){
      //your flyway init configuration...
    }
}
```
The trick here is to force FlywayMigrationInitializer to get instantiated after the ApplicationContextAware workaround

**Comment by zinefer** [comment]
I will try that when I get home. Thanks so much.

**Comment by purocean** [comment]
+1

**Comment by fladulima** [comment]
+1

**Comment by idallas456** [comment]
+1, this should really get fixed at some point.

**Comment by purocean** [comment]
Now I use my simple `Migration for app, very simple. I leave `Flayway` for long time.

https://github.com/purocean/kotlin-db-migration

1. simple & light
1. name based version
1. support dry run
1. support code migration and sql migration

**Comment by axelfontaine** [comment]
This has now been implemented generically using a new `javaMigrations` property in the API.

Spring users can use this to automatically use all `JavaMigration` Spring beans with Flyway:

```java
import org.flywaydb.core.Flyway;
import org.flywaydb.core.api.migration.JavaMigration;
import org.springframework.context.ApplicationContext;

...
ApplicationContext applicationContext = ...; // obtain a reference to Spring's ApplicationContext.

Flyway flyway = Flyway.configure()
    .dataSource(url, user, password)
    // Add all Spring-instantiated JavaMigration beans
    .javaMigrations(applicationContext.getBeansOfType(JavaMigration.class).values().toArray(new JavaMigration[0]))
    .load();
flyway.migrate();
```

**Comment by guylabs** [comment]
@axelfontaine Thanks for incorporating the feedback! I will have a look at it when I'll work again with a  Spring project. Maybe someones else in the thread is able to use it the way you implemented it. Always nice to see that feedback gets integrated!

**Comment by simonzhong1985** [comment]
> Sry I forgot to mention our configuration:
> 
> ```java
> @Configuration
> class FlywayConfiguration {
>     @Primary
>     @Bean(name = "flywayInitializer")
>     @DependsOn("springUtility") //or whatever your ApplicationContextAware is named
>     FlywayMigrationInitializer flywayMigrationInitializer(){
>       //your flyway init configuration...
>     }
> }
> ```
> 
> The trick here is to force FlywayMigrationInitializer to get instantiated after the ApplicationContextAware workaround



> This has now been implemented generically using a new `javaMigrations` property in the API.
> 
> Spring users can use this to automatically use all `JavaMigration` Spring beans with Flyway:
> 
> ```java
> import org.flywaydb.core.Flyway;
> import org.flywaydb.core.api.migration.JavaMigration;
> import org.springframework.context.ApplicationContext;
> 
> ...
> ApplicationContext applicationContext = ...; // obtain a reference to Spring's ApplicationContext.
> 
> Flyway flyway = Flyway.configure()
>     .dataSource(url, user, password)
>     // Add all Spring-instantiated JavaMigration beans
>     .javaMigrations(applicationContext.getBeansOfType(JavaMigration.class).values().toArray(new JavaMigration[0]))
>     .load();
> flyway.migrate();
> ```

@axelfontaine 
Could you help to provide one example to show how to inject spring bean into class implementation of JavaMigration interface?

Thanks,
Simon

**Comment by fitzyjoe** [comment]
@simonzhong1985 Right now all of my Flyway config is handled through the application.yml.  Can I still configure it that way when using your code - I'm referring to this comment:

//your flyway init configuration...

Also, I don't know what class the second snippet is in.  This is the snippet where you get a reference to Spring's ApplicationContext and call flyway.migrate().  What calls this code?

**Comment by jamesdh** [comment]
Has something changed since @axelfontaine's [comment](https://github.com/flyway/flyway/issues/1062#issuecomment-446279678)? It appears that having a JavaMigration as a Spring Component (and making use of dependency injection, etc) is automatically loaded and migrated without needing to define a `FlywayConfigurationCustomizer` that performs the example code he provided. Simply having the Bean-based JavaMigration in your regular source code (so Spring can scan it) as opposed to the Flyway migration location appears to be sufficient now.

**Comment by nbabb** [comment]
As @jamesdh pointed out, we are able to use BaseJavaMigration as spring @Component's, along with dependency injection, without the code cited above.  The migrations just need to be component scanned.

Once we upgraded to spring boot 2.7.4 however, we did have to set allow-circular-references=true to avoid circular dependency errors.  I've been trying to find a way to preserve what we have, without setting allow-circular-references=true, as that seems to be framed as an option of last resort.  I've yet to find a satisfactory solution.  

The command line runner strategy mentioned [here](https://stackoverflow.com/questions/60173562/spring-autowiring-in-flyway) allows the spring boot app to become accessible before the migrations are finished running, and has the added disadvantage of not being configurable via application.yml.  I've been searching for a way to trigger the migration at a point where we'd avoid circular dependency issues, but prior to the application becoming available, but haven't found a solution yet.

And the SpringUtility strategy mentioned [here](https://stackoverflow.com/questions/48214783/flyway-spring-boot-autowired-beans-with-jpa-dependency/48242865#48242865) which would allow us to keep the application.yml based configuration, didn't resolve the circular dependency issues for us.

We recognize that having migrations use repository/entity code is an issue when our object model changes.  The strategy we employ is to simply remove the migration body once this becomes a problem, and to create a new migration, if it is behavior we need to preserve.  We typically use java/spring based migrations for data seeding/migration tasks and not to create db structure.

It'd be nice if the ability to use spring/java based migrations (making full use of the repository layer) was documented and fully supported in the future.  Using JDBCTemplate is not a satisfactory solution for use, when we have a nice repository layer to work with.

**Comment by RohanBizongo** [comment]
Thanks @jamesdh  for the detailed answer. Helped me a lot. Would like to point out that we have been using Spring Boot version 2.7.16 and have not faced the crossroad of setting allow-circular-references to true. Annotating the migration class as a Component sufficed.

**Comment by la3rence** [comment]
If we set up a configuration bean by ourselves, we might exclude the `FlywayAutoConfiguration.class` with the annotation `@SpringBootApplication`, which will help to avoid the circular dependency error. See my [blog post]( https://lawrenceli.me/en/blog/flyway) for details that how to integrate Flyway and do DI with Springboot.

**Comment by Leon116** [comment]
Are there any news on this topic? I've created a [minimal example](https://github.com/Leon116/flyway-experiments), which shows that there is still a circular dependency issue when using Spring's DI mechanism in Flyway Java-based migrations.
However I found out that the problem only occurs when injecting JPA-related Spring Beans or Spring Beans having JPA-related beans in the bean dependency tree. If you are only injecting Spring Beans that have no association to JPA, having Java-based migrations as Spring Beans works just fine.
Implementing the solution suggested by @nbabb fixes the problem, but it just doesn't feel right having this nice JPA layer not working out of the box in database migrations.
 
**EDIT:** I opened an issue at spring-data-jpa. Feels like the problem maybe should be addressed there. See https://github.com/spring-projects/spring-data-jpa/issues/3959
